# Unique-Transcript-Finder

This pipeline works by constructing a non-redundant set of transcripts using GFFCompare. The number of times a transcript is found in all individuals is then counted and added to this database.
The database can then be queried with a transcript file in order to annotate for each transcript how many times it was found previously.
Then this annotated transcript file can be filtered according to personal criteria.
# Run
## Step 1: Constructing the database
The command to construct the database is:
```
__main__.py build <input> <gffcompare_container> <reference_transcriptome> [--optional_arguments]
```

Optional parameters are:
Optional argument | Function | Default
---|---|---
--database_name | Change the name of the file that will contain the database | transcript_database.gtf
--min_len | minimum length of transcripts to be included in the database | 50
--min_cov | minimum coverage of transcripts to be included in the database | 0
--min_fpkm | minimum fpkm of transcripts to be included in the database | 0
--min_tpm | minimum TPM of transcripts to be included in the database | 0
--min_iso | minimum isoform frequency of transcripts to be included in the database | 0.0
--out_prefix | out prefix for files generated by gff compare when counting transcript occurence in database | transcript_database
--target | Path where output files should be stored. Must end with \. If left empty files will be created in the current folder | 

For more information about running this command use
 ```
 python __main__.py build --help 
 ```
## Step 2: Querying the database
The command to query the database is:
```
__main__.py query <gff file> <gffcompare_container> [--optional_arguments]
```

Optional parameters are:
Optional argument | Function | Default
---|---|---
--database | Path to the transcript database created in step 1 | transcript_database.gtf
--database_pickle | Path to the pickle file containing count and frequency information of the database created in step 1 | transcript_database.gtf.pickle
--create_database_pickle | If the pickle file created in step 1 is lost, this option can be used to recreate the pickle file without having to rerun step 1 | False
--min_fpkm | minimum fpkm of transcripts to be included in the database | 0
--min_tpm | minimum TPM of transcripts to be included in the database | 0
--min_iso | minimum isoform frequency of transcripts to be included in the database | 0.0
--out_prefix | path to the GTF file with patient transcript information | query
--target_dir | Path where output files should be stored. Must end with \. If left empty files will be created in the current folder | 

For more information about running this command use
 ```
 python __main__.py query --help 
 ```
 
 ## Step 3: Filtering the annotated transcript files
 
 The command to query the database is:
```
__main__.py filter <annotated gff file> [--optional_arguments]
```

Optional parameters are:
Optional argument | Function | Default
---|---|---
--max_count | The max number of times a transcript can be found in the database to pass the filter. | 1
--max_frequency | The max frequecy at which a transcript can be included in the database to pass the filter. If this is changed it is adviced to set max_count arbitrarily high so that the count does not influence filtering| 1.0
--min_cov |The min coverage a transcript should have to pass the filter | 0.0
--min_fpkm | The min FPKM a transcript should have to pass the filter | 0.0
--min_tpm | The min TPM a transcript should have to pass the filter | 0.0
--annotated | If True, only transcript that are annotated in the reference genome pass the filter | True
--output_folder | Path where output files should be stored. Must end with \. If left empty files will be created in the current folder | 

For more information about running this command use
 ```
 python __main__.py filter --help 
 ```
 
 ## Extra option: statistics
 To obtain extra information about the database and transcripts in it run:
 ```
__main__.py statistics <rare threshold> <common_threshold> [--optional_arguments]
```

Optional parameters are:
Optional argument | Function | Default
---|---|---
--out_prefix | The outprefix used in step 1 | transcript_database
--target_folder | Path to the folder where all files were created during database building. Must end with \. If left empty files will be created in the current folder | 

For more information about running this command use
 ```
 python __main__.py statistics --help 
 ```
 
 
